Feature - provide seperation between concepts of `message pump` & `message handler`

stijnmoreels opened the pull request:
Relates to #35

tomkerkhove commented on src/Arcus.Messaging.Tests.Workers.ServiceBus.Queue/Program.cs:
```diff
@@ -36,7 +37,9 @@ public static void Main(string[] args)
                             .Build();
                     });
-                    services.AddServiceBusQueueMessagePump<OrdersMessagePump>(configuration => configuration["ARCUS_SERVICEBUS_CONNECTIONSTRING"]);
+                    services.AddServiceBusQueueMessagePump<OrdersMessagePump>(configuration => configuration["ARCUS_SERVICEBUS_CONNECTIONSTRING"])
```
I presume we don't need the message pump type here anymore?
```suggestion
                    services.AddServiceBusQueueMessagePump(configuration => configuration["ARCUS_SERVICEBUS_CONNECTIONSTRING"])
```

stijnmoreels replied:
No, we _do_ need that for the registration of the message pump.
The other calls registers the handlers.

tomkerkhove replied:
We need the Service Bus message pump, but that's it, the message handler is implementation specific.

tomkerkhove replied:
That's the goal of #35

stijnmoreels replied:
Yes, I think I get the goal, only not sure how this relates to this.
There's nothing changed on the way the message pump is added/registered, only the handlers are now processing the messages instead of the message pump.
So, we need to register both the pump and the handlers.

tomkerkhove replied:
What I expected to see would be:
```csharp
services.AddServiceBusQueueMessagePump(configuration => configuration["ARCUS_SERVICEBUS_CONNECTIONSTRING"])
        .WithMessageHandler<Order, OrdersMessageHandler>();
```

What would the `messageContext => true` be for? Filtering? If so, I'd leave that out for now but that's my personal opinion.

Something we should provide is a `WithFallbackMessageHandler` in case the above flow doesn't find a matching handler, but that might be more interesting for later.

tomkerkhove replied:
That way we have:
1) A message pump which checks the entity for new messages and triggers corresponding message handler
2) A message handler who takes action on the deserialized message

tomkerkhove commented on src/Arcus.Messaging.Tests.Workers/MessageHandlers/OrdersMessagePump.cs:
```diff
@@ -14,32 +14,12 @@ namespace Arcus.Messaging.Tests.Workers.MessageHandlers
 {
     public class OrdersMessagePump : AzureServiceBusMessagePump<Order>
```
This should no longer be required

tomkerkhove commented on src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs:
```diff
@@ -18,7 +18,7 @@ namespace Arcus.Messaging.Pumps.ServiceBus
     /// </summary>
     /// <typeparam name="TMessage">Type of expected message payload</typeparam>
-    public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessage, AzureServiceBusMessageContext>
+    public class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessage, AzureServiceBusMessageContext>
```
`TMessage` should no longer be required since that's what the `IMessageHandler` is for.

`MessagePump` should create an instance of `IMessageHandler` and hand it over to it

stijnmoreels replied:
So, without first deserializing?

tomkerkhove replied:
Ideally, the pump deserializes and passes it on to the message handler based on the type.

Maybe better to do a meeting on this?

stijnmoreels replied:
Updated this, can you take a look if it's something you had in mind?

tomkerkhove commented on src/Arcus.Messaging.Tests.Unit/ServiceBus/EmptyMessagePump.cs:
```diff
@@ -9,7 +9,7 @@
 namespace Arcus.Messaging.Tests.Unit.ServiceBus
 {
-    public class EmptyMessagePump : AzureServiceBusMessagePump<Order>
+    public class EmptyMessagePump : AzureServiceBusMessagePump
```
In theory you don't need this anymore

tomkerkhove commented on src/Arcus.Messaging.Tests.Workers.ServiceBus.Queue/Program.cs:
```diff
@@ -36,7 +36,9 @@ public static void Main(string[] args)
                             .Build();
                     });
-                    services.AddServiceBusQueueMessagePump<OrdersMessagePump>(configuration => configuration["ARCUS_SERVICEBUS_CONNECTIONSTRING"]);
+                    services.AddServiceBusQueueMessagePump(configuration => configuration["ARCUS_SERVICEBUS_CONNECTIONSTRING"])
```
This syntax looks perfect 👌

tomkerkhove commented on src/Arcus.Messaging.Tests.Workers/MessageHandlers/OrdersMessageHandler.cs:
```diff
@@ -0,0 +1,75 @@
+            CancellationToken cancellationToken)
+        {
+            Encoding encoding = messageContext.GetMessageEncodingProperty();
+            if (message.TryReadBodyAsJson(encoding, out Order order))
```
This should not be the responsibility of the message handler, the message pump should already do this and pass the order itself

tomkerkhove commented on src/Arcus.Messaging.Pumps.ServiceBus/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -225,18 +257,16 @@ public static IServiceCollection AddServiceBusTopicMessagePump<TMessagePump>(thi
 
-        private static void AddServiceBusMessagePump<TMessagePump>(IServiceCollection services, string entityName = null, string subscriptionName = null, Func<IConfiguration, string> getConnectionStringFromConfigurationFunc = null, Func<ISecretProvider, Task<string>> getConnectionStringFromSecretFunc = null, Action<AzureServiceBusMessagePumpOptions> configureMessagePump = null)
-            where TMessagePump : class, IHostedService
+        private static void AddServiceBusMessagePump<TMessagePump>(IServiceCollection services, string entityName = null, string subscriptionName = null, Func<IConfiguration, string> getConnectionStringFromConfigurationFunc = null, Func<ISecretProvider, Task<string>> getConnectionStringFromSecretFunc = null, Action<AzureServiceBusMessagePumpOptions> configureMessagePump = null) where TMessagePump ...
```
Do we still need the typed pump in the extensions?
```suggestion
        private static void AddServiceBusMessagePump(IServiceCollection services, string entityName = null, string subscriptionName = null, Func<IConfiguration, string> getConnectionStringFromConfigurationFunc = null, Func<ISecretProvider, Task<string>> getConnectionStringFromSecretFunc = null, Action<AzureServiceBusMessagePumpOptions> configureMessagePump = null) where TMessagePump : class, IHostedService
```

stijnmoreels replied:
Right, is fixed with merging with `master`.

tomkerkhove commented on src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs:
```diff
@@ -83,62 +82,80 @@ protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
                 }
-                catch (Exception ex)
+
+                bool isProcessed = await TryProcessMessageAsync(messageHandler, message, messageContext, correlationInfo, cancellationToken);
```
This is tricky, as discussed we should:
1. Try to deserialize the message according to the expected contract of the handler
2. If it matches, only the first match should process it
3. Done

Otherwise we're at risk of running a lot of duplicate processing, which failed, but still can cause inconsistent state

stijnmoreels commented:
@tomkerkhove , reflection approach available here: https://github.com/stijnmoreels/arcus.messaging/pull/1

tomkerkhove commented on src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs:
```diff
@@ -82,63 +81,71 @@ protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
             return Task.CompletedTask;
         }
 
+        // TODO: move to extension?
```
That's also an option, both are fine for me.

stijnmoreels replied:
Done that.

tomkerkhove commented on src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs:
```diff
@@ -82,63 +81,71 @@ protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
+            return DefaultEncoding;
         }
 
+        // TODO: should we also provide a 'streamed' version to process the message? We would require some way to reset the stream though (i.e. VirtualStream).
```
Let's hold this off for now maybe?

stijnmoreels replied:
Yes, of course.

tomkerkhove commented on src/Arcus.Messaging.ServiceBus.Core/Extensions/MessageExtensions.cs:
```diff
@@ -70,5 +72,70 @@ public static string GetTransactionId(this Message message)
+        /// <returns>
+        ///     Whether or not the deserialization succeeded [true], or not [false].
+        /// </returns>
+        public static bool TryReadBodyAsJson<TMessage>(this Message message, out TMessage output) where TMessage : class
```
Maybe good to add some unit tests for these new methods? They are pretty important.

stijnmoreels replied:
Yeah, removed that. Leftover from legacy approach.

tomkerkhove commented on src/Arcus.Messaging.Tests.Workers/MessageHandlers/OrdersMessageHandler.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Arcus.Messaging.Tests.Workers.MessageHandlers
+{
+    public class OrdersMessageHandler : IAzureServiceBusMessageHandler<Order>
```
Thinking out loud - If I were to not care about the Service Bus message context, can I still use `IMessageHandler<Order>`?

That would be very good to have imo. Reduces lockin.

tomkerkhove replied:
For clarity, I see value for both scenarios

stijnmoreels replied:
Well, the `IMessageHandler` uses the context in it's signature. So... we would have to create an different `IMessageHandler` interface with maybe an "ignored" message context so we can also include that in the filter.
When processing the message, the context would then be empty. or a `null`-type.

stijnmoreels replied:
In any case; different PR. Can you make an issue for this?

tomkerkhove replied:
See other comment, message context will always be there but it will just be specific to SB or the generic model.

tomkerkhove commented on src/Arcus.Messaging.Pumps.ServiceBus/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -246,5 +277,19 @@ private static void AddServiceBusMessagePump<TMessagePump>(IServiceCollection se
+        /// <typeparam name="TMessage"></typeparam>
+        /// <param name="services">The collection of services to use in the application.</param>
+        public static IServiceCollection WithMessagePumpHandler<TMessageHandler, TMessage>(this IServiceCollection services)
+            where TMessageHandler : class, IAzureServiceBusMessageHandler<TMessage>
```
This should not enforce `IAzureServiceBusMessageHandler` but `IMessageHandler` instead, otherwise there is no point in the decouple.

Imagine we add event hubs, I should have the option to use the specific interface for that one or use `IMessageHandler `but with less specific integration such as message context.

stijnmoreels replied:
Yes, right, but for the `AzureMessagePump`, that message pump can only use message handlers with a specific context. So, for that pump, we should only register those kind of implementations.

Maybe we should even change the signature to reflect this. `.WithAzureServiceBusMessageHandler`.

stijnmoreels replied:
We could maybe provide both. Since indeed, we could also register message handlers in a different context. But those will not be selected when processing messages in a Service Bus message pump.
That extension should probably be in another namespace too.

tomkerkhove replied:
Question is, why shouldn't we be allowed to use the generic message handlers? if I don't need the service bus message context and regular context is fine, it should be possible, no?

tomkerkhove commented on src/Arcus.Messaging.Pumps.ServiceBus/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -197,6 +213,24 @@ public static IServiceCollection AddServiceBusTopicMessagePump<TMessagePump>(thi
+        }
+
         /// <summary>
         ///     Adds a message handler to consume messages from Azure Service Bus Topic
```
We should update all the `AddServiceBusTopicMessagePump` & `AddServiceBusQueueMessagePump` extensions:

```suggestion
        ///     Adds a message pump to consume messages from Azure Service Bus Topic
```

message handlers are what we add via WithMessagePumpHandler

stijnmoreels replied:
Yes

tomkerkhove commented on src/Arcus.Messaging.Pumps.ServiceBus/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -246,5 +277,19 @@ private static void AddServiceBusMessagePump<TMessagePump>(IServiceCollection se
         }
+
+        /// <summary>
+        ///     Adds a <see cref="IAzureServiceBusMessageHandler{TMessage}"/> implementation to process the messages from Azure Service Bus resources.
```
Shouldn't be specific for Service Bus
```suggestion
        ///     Adds a message handler to process the messages.
```

stijnmoreels replied:
See comment above... this message pump can only handle those kind of message/contexts. We should maybe change the signature.

stijnmoreels replied:
Added a different extension for that.

tomkerkhove commented on src/Arcus.Messaging.Pumps.Abstractions/MessageHandling/MessageHandler.cs:
```diff
@@ -0,0 +1,203 @@
+            {
+                var serviceType = (Type) lookup.GetPropertyValue("Key");
+                
+                // TODO: update to more strict message handler filter.
```
Let's add an issue and remove the TODO

stijnmoreels replied:
Done: https://github.com/arcus-azure/arcus.messaging/issues/68

tomkerkhove commented on src/Arcus.Messaging.Pumps.Abstractions/MessageHandling/MessageHandler.cs:
```diff
@@ -0,0 +1,203 @@
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Arcus.Messaging.Pumps.Abstractions.MessageHandling
```
Can we provide tests for our voodoo part? ⚔😵

stijnmoreels replied:
We should really test the outside of this. Like, registering several message handlers with different messages and see if we can select them all.

tomkerkhove commented on src/Arcus.Messaging.ServiceBus.Core/Extensions/MessageContextExtensions.cs:
```diff
@@ -0,0 +1,35 @@
+                }
+                catch
+                {
+                    // Ignore.
```
Thought - Should we log it? Because something was specified but we couldn't parse it so we'll use UTF 8 instead.

Migth be easier for troubleshooting

stijnmoreels replied:
Hmm, then we have to pass in the logger.

tomkerkhove replied:
I wouldn't make it optional, otherwise you are in the dark

tomkerkhove replied:
Nevermind, it's ok

stijnmoreels commented:
> Question is, why shouldn't we be allowed to use the generic message handlers? if I don't need the service bus message context and regular context is fine, it should be possible, no?

I added another extension method to add just message handlers. The `AzureServiceMessagePump` only can handle handlers in its context, right? Or should we abbandon the context and just send the message with a `null` context instead then?

tomkerkhove commented:
> I added another extension method to add just message handlers. The `AzureServiceMessagePump` only can handle handlers in its context, right? Or should we abbandon the context and just send the message with a `null` context instead then?

It's simple: 
- When I use `IMessageHandler` I will only have `MessageContext`, regardless of the message pump that is being used
- When I use `IServiceBusMessageHandler` I'm explicitly pinning to that broker so I will get `ServiceBusMessagContext`

stijnmoreels commented:
> > I added another extension method to add just message handlers. The `AzureServiceMessagePump` only can handle handlers in its context, right? Or should we abbandon the context and just send the message with a `null` context instead then?
> 
> It's simple:
> 
> * When I use `IMessageHandler` I will only have `MessageContext`, regardless of the message pump that is being used
> * When I use `IServiceBusMessageHandler` I'm explicitly pinning to that broker so I will get `ServiceBusMessagContext`

Updated this with another shortcut message handler.

stijnmoreels commented:
> LGTM - Do you want to do docs here or seperate PR?

Yes, I think that would be best.
